File: tests/rosetta/transpiler/php/cullen-and-woodall-numbers.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
function _iadd($a, $b) {
    if (function_exists('bcadd')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcadd($sa, $sb, 0);
    }
    return $a + $b;
}
function _isub($a, $b) {
    if (function_exists('bcsub')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcsub($sa, $sb, 0);
    }
    return $a - $b;
}
function _imul($a, $b) {
    if (function_exists('bcmul')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcmul($sa, $sb, 0);
    }
    return $a * $b;
}
function _idiv($a, $b) {
    return _intdiv($a, $b);
}
function _imod($a, $b) {
    if (function_exists('bcmod')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcmod($sa, $sb));
    }
    return $a % $b;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function pow_big($base, $exp) {
  $result = 1;
  $b = $base;
  $e = $exp;
  while ($e > 0) {
  if ($e % 2 == 1) {
  $result = _imul($result, $b);
}
  $b = _imul($b, $b);
  $e = intval((_intdiv($e, 2)));
};
  return $result;
};
  function cullen($n) {
  $two_n = pow_big(2, $n);
  return _iadd((_imul($two_n, ($n))), (1));
};
  function woodall($n) {
  return _isub(cullen($n), (2));
};
  function show_list($xs) {
  $line = '';
  $i = 0;
  while ($i < count($xs)) {
  $line = $line . _str($xs[$i]);
  if ($i < count($xs) - 1) {
  $line = $line . ' ';
}
  $i = $i + 1;
};
  return $line;
};
  function main() {
  $cnums = [];
  $i = 1;
  while ($i <= 20) {
  $cnums = array_merge($cnums, [cullen($i)]);
  $i = $i + 1;
};
  echo rtrim('First 20 Cullen numbers (n * 2^n + 1):'), PHP_EOL;
  echo rtrim(show_list($cnums)), PHP_EOL;
  $wnums = [];
  $i = 1;
  while ($i <= 20) {
  $wnums = array_merge($wnums, [woodall($i)]);
  $i = $i + 1;
};
  echo rtrim('
First 20 Woodall numbers (n * 2^n - 1):'), PHP_EOL;
  echo rtrim(show_list($wnums)), PHP_EOL;
  $cprimes = [1, 141, 4713, 5795, 6611];
  echo rtrim('
First 5 Cullen primes (in terms of n):'), PHP_EOL;
  echo rtrim(show_list($cprimes)), PHP_EOL;
  $wprimes = [2, 3, 6, 30, 75, 81, 115, 123, 249, 362, 384, 462];
  echo rtrim('
First 12 Woodall primes (in terms of n):'), PHP_EOL;
  echo rtrim(show_list($wprimes)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/rosetta/transpiler/FS/cullen-and-woodall-numbers.fs
// Generated 2025-07-30 21:41 +0700

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let rec pow_big (``base``: bigint) (exp: int) =
    let mutable __ret : bigint = Unchecked.defaultof<bigint>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    try
        let mutable result: bigint = bigint 1
        let mutable b: bigint = ``base``
        let mutable e: int = exp
        while e > 0 do
            if (((e % 2 + 2) % 2)) = 1 then
                result <- result * b
            b <- b * b
            e <- int (e / 2)
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and cullen (n: int) =
    let mutable __ret : bigint = Unchecked.defaultof<bigint>
    let mutable n = n
    try
        let two_n: bigint = pow_big (bigint 2) n
        __ret <- (two_n * (bigint n)) + (bigint 1)
        raise Return
        __ret
    with
        | Return -> __ret
and woodall (n: int) =
    let mutable __ret : bigint = Unchecked.defaultof<bigint>
    let mutable n = n
    try
        __ret <- (cullen n) - (bigint 2)
        raise Return
        __ret
    with
        | Return -> __ret
and show_list (xs: bigint array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable xs = xs
    try
        let mutable line: string = ""
        let mutable i: int = 0
        while i < (Seq.length xs) do
            line <- line + (string (xs.[i]))
            if i < ((Seq.length xs) - 1) then
                line <- line + " "
            i <- i + 1
        __ret <- line
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let mutable cnums: bigint array = [||]
        let mutable i: int = 1
        while i <= 20 do
            cnums <- Array.append cnums [|unbox<bigint> (cullen i)|]
            i <- i + 1
        printfn "%s" "First 20 Cullen numbers (n * 2^n + 1):"
        printfn "%s" (show_list cnums)
        let mutable wnums: bigint array = [||]
        i <- 1
        while i <= 20 do
            wnums <- Array.append wnums [|unbox<bigint> (woodall i)|]
            i <- i + 1
        printfn "%s" "\nFirst 20 Woodall numbers (n * 2^n - 1):"
        printfn "%s" (show_list wnums)
        let cprimes: bigint array = [|bigint 1; bigint 141; bigint 4713; bigint 5795; bigint 6611|]
        printfn "%s" "\nFirst 5 Cullen primes (in terms of n):"
        printfn "%s" (show_list cprimes)
        let wprimes: bigint array = [|bigint 2; bigint 3; bigint 6; bigint 30; bigint 75; bigint 81; bigint 115; bigint 123; bigint 249; bigint 362; bigint 384; bigint 462|]
        printfn "%s" "\nFirst 12 Woodall primes (in terms of n):"
        printfn "%s" (show_list wprimes)
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
